### PR TITLE
fixes scroll restoration after native navigation to another sveltekit route return

### LINKED ISSUE #4386: history scroll restoration usage with native navigation
### Describe the problem

I first noticed this when I was browsing around a website built with Next.js, and since #3873 and #3938 we're using similar technique for restoring scroll position as Next, so this bug applies too, but in much smaller scale since we listen to all anchor clicks, and not just Link component like Next does.

The issue is when the page does native navigation to another link that is being handled by SvelteKit, and that second navigation overwrites scroll position inside session storage, so when user returns to the first page, second scroll position is being read.

This is pretty edge case for SvelteKit because of listening to all anchor elements, but still it could happen.

### Describe the proposed solution

I can push PR for this.

### Alternatives considered

Leave things the way they currently are. Since this is edge case that could be okay.

### Importance

nice to have

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by PH4NTOMiki**
### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [ ] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpx changeset` and following the prompts. All changesets should be `patch` until SvelteKit 1.0

Closes #4386
Probably needs test.

### CODE REVIEW

**Comment by PatrickG** [comment]
I was about to look into this issue as well.
I think it would be far easier to change `scroll_positions` into an array and set the `.length` property to `current_history_index + 1` when clicking a link or using `goto`.

**Comment by PH4NTOMiki** [comment]
@Rich-Harris what do you think about this?

**Comment by PH4NTOMiki** [comment]
@PatrickG I don't think that what you suggested would solve #4386 because we need to make collisions impossible, and that won't do it, this will because it uses random keys.

**Comment by PH4NTOMiki** [comment]
And in case I wasn't really clear in the issue, the issue goes like this: 

1. go to some Kit app (the `history.state` would be like this: `{sveltekit:index: 0}`) and scroll to the end/bottom of the page
2. you click on some internal link in that app and for some reason(maybe `router=false`, maybe `app version changed` etc.) navigation is handled by the `browser native` navigation
3. after that the `history.state` would again be `{sveltekit:index: 0}` and that would cause overwrite of position state.
4. go back using browser back button, you would't be returned to the bottom of the page where scroll position should be because you left it there in step 1. because it was overwritten by the second navigation.

**Comment by PatrickG** [comment]
> And in case I wasn't really clear in the issue, the issue goes like this:
> 
> 1. go to some Kit app (the `history.state` would be like this: `{sveltekit:index: 0}`) and scroll to the end/bottom of the page
> 2. you click on some internal link in that app and for some reason(maybe `router=false`, maybe `app version changed` etc.) navigation is handled by the `browser native` navigation
> 3. after that the `history.state` would again be `{sveltekit:index: 0}` and that would cause overwrite of position state.
> 4. go back using browser back button, you would't be returned to the bottom of the page where scroll position should be because you left it there in step 1. because it was overwritten by the second navigation.

Alright, then i misunderstood the issue.
I thought the issue is:
- start page (index 0): click link -> second page (index 1)
- scroll down on second page (index 1)
- go back to start page (index 0)
- klick another link -> thrid page (index 1) -> page scrolls down unintentionally

**Comment by benmccann** [comment]
I agree we should match the browser behavior and not share scroll state across tabs. FYI, Rich is on vacation this week

File: packages/kit/src/runtime/client/client.js [hunk 1]
```diff
@@ -33,15 +34,15 @@ const default_error = components[1]();
 // state we're navigating from
 
 /** @typedef {{ x: number, y: number }} ScrollPosition */
-/** @type {Record<number, ScrollPosition>} */
+/** @type {Record<String, ScrollPosition>} */
```
Comment (benmccann) [inline packages/kit/src/runtime/client/client.js:37]: We normally use lowercase `string` and `number`

https://stackoverflow.com/a/42654496

File: packages/kit/src/runtime/client/client.js [hunk 2]
```diff
@@ -259,6 +269,7 @@ export function create_client({ target, session, base, trailing_slash }) {
 			const { details } = opts;
 			const change = details.replaceState ? 0 : 1;
 			details.state[INDEX_KEY] = current_history_index += change;
+			details.state[RESTORATION_KEY] = current_position_index = Math.random().toString(32).slice(2);
```
Comment (PH4NTOMiki) [inline packages/kit/src/runtime/client/client.js:272]: I'm not sure whether I should put this in `if(!details.replaceState)` contitional

File: packages/kit/src/runtime/client/client.js [hunk 3]
```diff
@@ -19,6 +19,7 @@ import { components, dictionary, matchers } from '__GENERATED__/client-manifest.
 
 const SCROLL_KEY = 'sveltekit:scroll';
 const INDEX_KEY = 'sveltekit:index';
+const RESTORATION_KEY = 'sveltekit:restoration';
```
Comment (PH4NTOMiki) [inline packages/kit/src/runtime/client/client.js:22]: And I'm not sure about this name.

File: packages/kit/src/runtime/client/client.js (hunk 2 above)
Comment (PH4NTOMiki) [reply packages/kit/src/runtime/client/client.js]: I did put in inside conditional

File: packages/kit/src/runtime/client/client.js (hunk 3 above)
Comment (PH4NTOMiki) [reply packages/kit/src/runtime/client/client.js:23]: Still not sure, not changing for now

**Comment by PH4NTOMiki** [comment]
> Your solution does fix that as well.

I thought that was working already, do you have a example where it wasn't?

**Comment by PH4NTOMiki** [comment]
@benmccann is it okay to merge this, it works locally, I tested it and it works, and it's really difficult to add test for this because it's pretty edge-case-y.

**Comment by PH4NTOMiki** [comment]
@Rich-Harris can you check this out?

**Comment by PH4NTOMiki** [comment]
@Rich-Harris can you merge this, or just say why it's not merged.

**Comment by Rich-Harris** [comment]
Can you add a test?

**Comment by PH4NTOMiki** [comment]
@Rich-Harris It's pretty hard thing to test for, because the bug happens in some specific conditions, that are hard to create in testing environment. But if you check out the code, you can see that there is not any way that this code could break anything, and I checked out the deployment https://kit-svelte-dev-zu82n3ukp-svelte.vercel.app/ and it works in 3 of my browsers(Edge, Firefox, Chrome)

**Comment by PH4NTOMiki** [comment]
@Rich-Harris but I can hope we can merge it anyway(without the test)

**Comment by Rich-Harris** [comment]
I haven't managed to write a test that fails on `master`, but I _was_ able to simply the implementation somewhat. There's no need to maintain `SCROLL_INDEX_KEY` separately — all we need to do is ensure that cross-document navigations within the same app (or between two SvelteKit apps) don't corrupt the scroll position history, which is easy to do by offsetting the history index. Closing this in favour of #4640 — thanks
